system/classes/KO7/Controller/REST.php

```diff
@@ -0,0 +1,134 @@
+<?php
+/**
+ * Abstract Controller class for REST controller mapping.
+ * Supports GET, PUT, POST, and DELETE.
+ *
+ * @package        KO7\REST
+ *
+ * @copyright  (c) since 2016 Koseven Team
+ * @license        https://koseven.ga/LICENSE
+ */
+abstract class KO7_Controller_REST extends Controller {
+
+    /**
+     * Default Output (only if allowed by configuration)
+     *
+     * @var string
+     */
+    public static $default_output = 'json';
+
+    /**
+     * REST types
+     *
+     * @var array
+     */
+    protected static $_action_map = [
+        HTTP_Request::GET => 'index',
+        HTTP_Request::PUT => 'update',
+        HTTP_Request::POST => 'create',
+        HTTP_Request::DELETE => 'delete'
+    ];
+
+    /**
+     * Automatically executed before the controller action.
+     * Evaluate Request (method, action, parameter, format)
+     */
+    public function before() : void
+    {
+        // Parent call
+        parent::before();
+
+        // Inject output format if not set by route
+        $this->request->inject_param('format', $this->request->param('format') ?: static::$default_output);
```

Comment:
1. $this->request->param() have second param `$default`
2. format can contains in param['format'], query['format'] or header['accept']
3. Method `inject_param` is wrong way, dont rewrite request data. Why u want contains format as request param?

~~~
// replace static $default_output to
protected $format = 'json';
...
if ($this->request->param('format')) {
 $this->format = $this->request->param('format');
} elseif ($this->request->query('format')) {
 $this->format = $this->request->query('format');
} elseif ($this->request->accept_type()) {
  list(, $this->format) = explode('/', $this->request->accept_type()[0]);
}
~~~

Reply:
1. Yes but does not allow to inject one
2. Not sure if we should let the format option decide via accept header.... i think a predefined one (either via routes or as default) is enough. Your code will not work if the user adds custom driver
3. For me it is because the format is part of the request (User requests data in xx format)....i am also a bit unsure if injecting is the right thing....maybe i find a better way to archive this.

Reply:
Removed the `inject_param` function, added a `$_format` property to request which can be recived or set via `format` method (like other variables in koseven).
Now the requested format can be found there and be easy accessed within the application, because as said the "**requested** format" belongs to me in the "**request**" class

For now the format is received via route parameter, i will check if setting via accept header will be implemented....